starlark/testdata/bytes.star

```diff
@@ -0,0 +1,131 @@
+assert.eq(b"goodbye", goodbye)
+assert.eq(b"", empty)
+assert.eq(b"\t\n\u200D", nonprinting)
+assert.ne("abc", b"abc")
+assert.eq(b"\012\xff\u0400\U0001F63F", b"\n\xffЀ😿") # see scanner tests for more
+assert.eq(rb"\r\n\t", b"\\r\\n\\t") # raw
+
+# type
+assert.eq(type(hello), "bytes")
+
+# len
+assert.eq(len(hello), 13)
+assert.eq(len(goodbye), 7)
+assert.eq(len(empty), 0)
+
+# truth
+assert.true(hello)
+assert.true(goodbye)
+assert.true(not empty)
+
+# str(bytes) does UTF-8 to UTF-k transcoding.
+assert.eq(str(hello), "hello, 世界")
+assert.eq(str(hello[:-1]), "hello, 世��")  # incomplete UTF-8 encoding => U+FFFD
+assert.eq(str(goodbye), "goodbye")
+assert.eq(str(empty), "")
+assert.eq(str(nonprinting), "\t\n\u200d")
+assert.eq(str(b"\udc00"), "�") # unpaired surrogate => U+FFFD
+
+# reprq
+assert.eq(repr(hello), r'b"hello, 世界"')
+assert.eq(repr(hello[:-1]), r'b"hello, 世\xe7\x95"')  # (incomplete UTF-8 encoding )
+assert.eq(repr(goodbye), 'b"goodbye"')
+assert.eq(repr(empty), 'b""')
+assert.eq(repr(nonprinting), 'b"\\t\\n\\u200d"')
+
+# equality
+assert.eq(hello, hello)
+assert.ne(hello, goodbye)
+assert.eq(bytes("goodbye"), goodbye)
+
+# ordered comparison
+assert.lt(bytes("abc"), bytes("abd"))
+assert.lt(bytes("abc"), bytes("abcd"))
+assert.lt(bytes("\x7f"), bytes("\x80")) # bytes compare as uint8, not int8
+
+# bytes are dict-hashable
+dict = {hello: 1, goodbye: 2}
+dict[bytes("goodbye")] = 3
+assert.eq(len(dict), 2)
+assert.eq(dict[goodbye], 3)
+
+# indexing
+assert.eq(goodbye[0], bytes("g"))
+assert.eq(goodbye[-1], bytes("e"))
+assert.fails(lambda: goodbye[100], "out of range")
+
+# slicing
+assert.eq(goodbye[:4], bytes("good"))
+assert.eq(goodbye[4:], bytes("bye"))
+assert.eq(goodbye[::2], bytes("gobe"))
+assert.eq(goodbye[3:4], bytes("d"))  # special case: len=1
+assert.eq(goodbye[4:4], bytes(""))  # special case: len=0
+
+# ord
+assert.eq(ord(b"a"), 97)
+assert.fails(lambda: ord(b"ab"), "ord: bytes has length 2, want 1")
+assert.fails(lambda: ord(b""), "ord: bytes has length 0, want 1")
+
+# repeat (bytes * int)
+assert.eq(goodbye * 3, bytes("goodbyegoodbyegoodbye"))
+assert.eq(3 * goodbye, bytes("goodbyegoodbyegoodbye"))
+
+# elems() returns an iterable value over 1-byte substrings.
+assert.eq(type(hello.elems()), "bytes.elems")
+assert.eq(str(hello.elems()), "b\"hello, 世界\".elems()")
+assert.eq(list(hello.elems()), [104, 101, 108, 108, 111, 44, 32, 228, 184, 150, 231, 149, 140])
+assert.eq(bytes([104, 101, 108, 108, 111, 44, 32, 228, 184, 150, 231, 149, 140]), hello)
+assert.eq(list(goodbye.elems()), [103, 111, 111, 100, 98, 121, 101])
+assert.eq(list(empty.elems()), [])
+
+# TODO(adonovan);
+# - more methods: find, index, split, etc.
+# - hash(bytes)?
+#
+# Summary of string operations (put this in spec).
+#
+# string to number:
+# - bytes[i]  returns numeric value of ith byte.
+# - ord(string)  returns numeric value of sole code point in string.
+# - ord(string[i])  is not a useful operation: fails on non-ASCII; see below.
```

Comment:
This seems unfortunate. I feel like I'm getting mixed signals about whether strings are unicode data or sequences of k-bit elements. If it's the latter, you'd think we'd be freer to manipulate and inspect their contents, including getting the integer value of an element. It's enough to make me think `ord()` should have a companion `codepoint_ord()` toplevel function.

Reply:
See my [comment](https://github.com/bazelbuild/starlark/issues/112#issuecomment-750357023) on the design thread for more on this issue and the one I mentioned below.